ENH: Initial implementation of rotation composition

adbugger opened the pull request:
It appears that `np.einsum` automatically handles all shape combinations for us. Will upload test cases next.

adbugger committed:
ENH: Initial implementation of rotation composition

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -9,6 +9,33 @@
+    num_quat = q.shape[0]
+    q_mat = np.empty((num_quat, 4, 4))
+    # w
+    q_mat[:, 0, 0] = q_mat[:, 1, 1] = q_mat[:, 2, 2] = q_mat[:, 3, 3] = q[:, 3]
```
I'm not sure that I like this in terms of readability (multiple assignments per line). I don't mind if you keep it that way.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -9,6 +9,33 @@
 AXIS_TO_IND = {'x': 0, 'y': 1, 'z': 2}
 
 
+def _rep_mat(q):
```
Think of a better name.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -629,3 +656,17 @@ def inv(self):
             quat = quat[0]
         return self.__class__(quat, normalized=True)
+
+    def __mul__(self, other):
```
What about the docstring? You can write it the same way and add this to Methods.

adbugger replied:
Ah ok, I didn't know a docstring was required for such methods.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -629,3 +656,17 @@ def inv(self):
         return self.__class__(quat, normalized=True)
+
+    def __mul__(self, other):
+        p = self._quat
```
Not sure what value this assignments bring.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -629,3 +656,17 @@ def inv(self):
+    def __mul__(self, other):
+        p = self._quat
+        q = other._quat
+        q_mat = _rep_mat(q)
```
I think we should convert the object with the smallest number of rotations in it (that is 1 when 1 x N is presented).

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -629,3 +656,17 @@ def inv(self):
+
+        # einsum handles Nx1, NxN, 1xN for us
+        # raises ValueError if broadcasting is not possible
+        result = np.einsum('...ij,...ijk->...ik', p, q_mat)
```
One thing we need to realize is that `np.einsum` is not equivalent to `np.dot` in terms of performance:

So for example:
```
import numpy as np
a = np.random.rand(4, 4)
b = np.random.rand(10000, 4)
%timeit b.dot(a.T)
%timeit np.einsum('...ij,...j->...i', a, b)
```
You will see that the dot approach works about 3 times faster. 

I think we can still use `einsum`.

adbugger replied:
Given that we want to expand the object with the smaller number of rotations, I think it's best to handle each case `Nx1`, `NxN` and `1xN` separately. I'll make the changes and push.

larsoner replied:
What gain is there by doing that? I don't think the 3x gain will be worth the additional code complexity.

adbugger replied:
In the case of `Nx1` and `NxN` compositions, we have to convert the second object to a matrix and multiply.

For the case of `1xN` composition, we should probably convert the first object to a matrix (converting the second one seems wasteful). Since the order is now reversed, this composition will involved transposing the matrices before multiplication. This case needs to be handled differently anyway.

If we're aggressively trying to stick to one approach, then we could go with simple quaternion composition across the board. It's slower, but there are no cases attached to that.

larsoner replied:
> It's slower, but there are no cases attached to that.

See my previous email about maintainability, numerical stability, and optimization -- I'd rather get clean and simple, more maintainable code working and merged, unless it's obvious that for expected realistic use cases there is a real benefit of adding code complexity. So if the quat composition isn't horribly slower (> 10x?) it seems fine to use it here. We can always optimize later when people hit limiting use cases (which might be never!).

Perhaps cleaner than this code would involve:

1. Allocating the output array based on input shape
2. Computing each output quat in a separate statement involving `[..., 0]`-style indexing.

Then I think you won't even need the `_single` check at the end because NumPy broadcasting will take care of this for you.

FWIW, if this is what you mean by "direct composition", I'm actually surprised that it's slower than what you wrote here. Both the `np.dot`-style operation and the "direct" method can make use of SIMD instructions for speed.

larsoner replied:
Actually I'm not convinced the idea I had will work, so feel free to ignore it. But I am curious what the alternative code would look like, I imagine it would be simpler (and the `_rep_mat` would go away?)

adbugger replied:
I've already uploaded the relevant benchmarks and code to the wiki [here](https://github.com/adbugger/scipy/wiki/Benchmarking-experiments-for-rotation-compositions).

For the `Nx1` case, the quaternion approach is ~4x slower until we reach astronomical numbers for `N`, where is becomes closer to ~8x slower.

For the `NxN` case, the quaternion approach is a little less then ~4x slower, but quickly starts outperforming the matrix approach at `~1e3` rotations.

Take a look and let me know.

nmayorov replied:
It's hard for me to understand how to approach optimization vs maintainability issues.

What Eric suggesting is to almost ignore all performance tricks. My problems with that it is unlikely to be changed afterwards, because other people won't get involved enough to figure it out. On the other hand, perhaps this "problem" is not that significant if it won't be obviously seen by others.

So if we adapt it, then I agree with Aditya:

> If we're aggressively trying to stick to one approach, then we could go with simple quaternion composition across the board. It's slower, but there are no cases attached to that.

I suggest to stick to this approach. It obviously has some advantages of being conceptually elegant and memory efficient.

-----------

As we are here, what about rotations applications? In this case we could again use quaternion formulation or get `dcm` by `as_dcm` and use dot product / einsum. In this case I'm even inclined to use `as_dcm` because it is already written and it is natural to apply transformation through matrix. (Also it will be quicker likely.).

nmayorov reviewed (commented):
Some thoughts. I'm not totally sure that we must stick with `einsum`, but it is simple in implementation.

adbugger committed:
MAINT: Simpler implementation

adbugger committed:
MAINT: Removed unused function

Since we're not using the matrix representation anymore, remove the
function responsible for generating that representation.

adbugger commented:
The `__mul__` method is not being documented even though I added it to the list of Methods. I looked at the code from `_cubic.py` from the `scipy.interpolate` package. They seem to be doing the same thing.

nmayorov commented:
The new code looks really nice, very simple. 

One thing - I slightly more prefer to check shapes match before doing the composition, and write some specific message if they don't match.

I will try to figure out the issue with the method not being documented.

Meanwhile start to work on rotation application.

adbugger committed:
MAINT: Added error message for shape mismatch.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -629,3 +629,31 @@ def inv(self):
+    def __mul__(self, other):
+        """Returns the composition of the rotations.
+
+        If `p` and `q` are two rotations, then the composition of 'q followed
```
It should be "p followed by q", right?

adbugger replied:
I checked this, and quaternions compose the same way as dcms, i.e. right to left. Check out slide 18 of [this](http://run.usc.edu/cs520-s12/quaternions/quaternions-cs520.pdf) presentation. It seems to corroborate my view. Of course, we could choose a convention for `p * q` to be `p` followed by `q`, which would amount to `_quat_compose(quat_q, quat_p)`.

nmayorov replied:
I'm confused, "p * q", means "p then q", means "p followed by q". I.e. it is q that follows p.

I'm not taking about convention changes, simply about language. Am I wrong in this part (English phrasing)?

adbugger replied:
Well, I think both of us got a little confused here.
When writing `p * q`, I'm using the same convention that dcms follow. So `p * q` is the same as "q then p", and "p follows q".

nmayorov replied:
I understand you now. You mean vector transformations, whereas I mean physical frame rotations, like first we rotate by p and then by q (to get the final frame).

Let's think about it some more, and then update the description.

adbugger replied:
Ah yes, of course. Both are equally valid ways of looking at it.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -629,3 +629,31 @@ def inv(self):
+        """Returns the composition of the rotations.
+
+        If `p` and `q` are two rotations, then the composition of 'q followed
+        by p' is equivalent to `p * q`, which is the same as
```
I suggest to rephrase it as: "In terms of DCM, the composition is expressed as ...". That is be slightly more clear, that we can consider the composition as dcm multiplication.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -629,3 +629,31 @@ def inv(self):
+
+            - Either `p` or `q` contains a single rotation. In this case the
+              returned object contains the result of composing each rotation in
+              the other object with the single rotation
```
Add period.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -629,3 +629,31 @@ def inv(self):
         return self.__class__(quat, normalized=True)
+
+    def __mul__(self, other):
+        """Returns the composition of the rotations.
```
PEP prescribes to write it as "Return the composition of the rotation". But perhaps write it as "Compose this rotation with other."

nmayorov reviewed (commented):
Some minor comments, you can address them in a separate commit(s). No pr needed.

nmayorov commented:
I figured why `__mul__` is not being documented. Check out this file https://raw.githubusercontent.com/scipy/scipy/master/doc/source/_templates/autosummary/class.rst

You see that `__call__` is explicitly mentioned there, so add `__mul__` to the list and it will work (tested). Fix it by a commit.
